x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlanner.java

```diff
@@ -0,0 +1,193 @@
+                        break;
+                    default:
+                        throw new IllegalStateException();
+                }
+            }
+        }
+
+        return new Node(n.id(), n.availableMemoryBytes() - bytesUsed, n.cores() - coresUsed);
+    }
+
+    private Model modifyModelPreservingPreviousAssignments(Model m, PreserveAllocations preserveAllocations) {
+        if (m.currentAllocationByNodeId().isEmpty()) {
+            return m;
+        }
+
+        int preservedAllocations = 0;
+        Map<String, Integer> remainingAllocationsToPreservePerNodeId = new HashMap<>();
+        switch (preserveAllocations) {
+            case ONE:
+                preservedAllocations += m.currentAllocationByNodeId().values().stream().filter(v -> v > 0).count();
+                m.currentAllocationByNodeId()
+                    .entrySet()
+                    .forEach(e -> remainingAllocationsToPreservePerNodeId.put(e.getKey(), Math.max(0, e.getValue() - 1)));
+                break;
+            case ALL:
+                preservedAllocations += m.currentAllocationByNodeId().values().stream().mapToInt(Integer::intValue).sum();
+                m.currentAllocationByNodeId().keySet().forEach(k -> remainingAllocationsToPreservePerNodeId.put(k, 0));
+                break;
+            default:
+                throw new IllegalStateException();
+        }
+        return new Model(
+            m.id(),
+            m.memoryBytes(),
+            m.allocations() - preservedAllocations,
+            m.threadsPerAllocation(),
+            remainingAllocationsToPreservePerNodeId
+        );
+    }
+
+    private AssignmentPlan mergePreviousAssignments(AssignmentPlan assignmentPlan, PreserveAllocations preserveAllocations) {
+        AssignmentPlan.Builder mergedPlanBuilder = AssignmentPlan.builder(nodes, models);
+        for (Model m : models) {
+            Map<Node, Integer> assignments = assignmentPlan.assignments(m);
+            for (Node n : nodes) {
+                int allocations = assignments == null ? 0 : assignments.getOrDefault(n, 0);
+                if (m.currentAllocationByNodeId().containsKey(n.id())) {
+                    switch (preserveAllocations) {
+                        case ONE:
+                            allocations++;
+                            break;
+                        case ALL:
+                            allocations += m.currentAllocationByNodeId().get(n.id());
+                            break;
+                        default:
+                            throw new IllegalStateException();
+                    }
+                }
+                if (allocations > 0) {
+                    mergedPlanBuilder.assignModelToNode(m, n, allocations);
+                }
+            }
+        }
+        return mergedPlanBuilder.build();
+    }
+
+    private enum PreserveAllocations {
+        ONE,
+        ALL;
+    }
```

Comment:
I think all this code would be simpler, and probably less susceptible to bugs (as it will reduce predicates on this down to 1), if we had a 

`AbstractPreserveAllocation` class and a `PreserveOneAllocation` and `PreserveAllAllocation` or something. 

Their constructors take the `List<Node> nodes, List<Model> models` and can calculate the required values. 

Then, we would have only one switch statement (constructing those subclasses).